usr/lib/common/mech_openssl.c

```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
+                                          &modulus);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_MODULUS\n");
+        goto done;
+    }
+
+    n = BN_CTX_get(bn_ctx);
+    if (n == NULL ||
+        BN_bin2bn(modulus->pValue, modulus->ulValueLen, n) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for modulus\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(n, BN_FLG_CONSTTIME);
+
+    /* Get public exponent a BIGNUM */
+    rc = template_attribute_get_non_empty(key_obj->template,
+                                          CKA_PUBLIC_EXPONENT, &pub_exp);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_PUBLIC_EXPONENT\n");
+        goto done;
+    }
+
+    e = BN_CTX_get(bn_ctx);
+    if (e == NULL ||
+        BN_bin2bn(pub_exp->pValue, pub_exp->ulValueLen, e) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for public exponent\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(e, BN_FLG_CONSTTIME);
+
+    /* Get prime1 a BIGNUM */
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_PRIME_1,
+                                          &prime1);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_PRIME_1\n");
+        goto done;
+    }
+
+    p = BN_CTX_get(bn_ctx);
+    if (p == NULL ||
+        BN_bin2bn(prime1->pValue, prime1->ulValueLen, p) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for prime1\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(p, BN_FLG_CONSTTIME);
+
+    /* Get prime2 a BIGNUM */
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_PRIME_2,
+                                          &prime2);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_PRIME_2\n");
+        goto done;
+    }
+
+    q = BN_CTX_get(bn_ctx);
+    if (q == NULL ||
+        BN_bin2bn(prime1->pValue, prime1->ulValueLen, q) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for prime2\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(q, BN_FLG_CONSTTIME);
+
+    d = BN_CTX_get(bn_ctx);
+    if (d == NULL) {
+        TRACE_ERROR("BN_CTX_get failed d\n");
+        rc =  CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(d, BN_FLG_CONSTTIME);
+
+    /*
+     * phi(n) = (p - 1 )(q - 1) = n - p - q + 1
+     * d = e ^{-1} mod phi(n).
+     */
+    if (BN_copy(d, n) == NULL ||
+        BN_sub(d, d, p) == 0 ||
+        BN_sub(d, d, q) == 0 ||
+        BN_add_word(d, 1) == 0 ||
+        BN_mod_inverse(d, e, d, bn_ctx) == NULL) {
+        TRACE_ERROR("Failed to calculate private key part d\n");
+        rc =  CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    if (BN_bn2binpad(d, priv_exp, priv_exp_len) <= 0) {
+        TRACE_ERROR("BN_bn2binpad failed\n");
+        rc =  CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+done:
+    BN_CTX_free(bn_ctx);
```

Comment:
Interesting: looks like there is no BN_CTX_clear_free, similar to BIGNUMs ...

Reply:
Thats the beauty of a BN-context. If you free the BN-context, it automatically frees all BIGNUMs that you obtained from it (via BN_CTX_get()). And BN_CTX_free() also does do a BN_CTX_clear_free() internally on all its BIGNUMs. 
BTW: This here is a secure BN-context (BN_CTX_secure_new()), so all BIGNUMs a resecure BIGNUMs.